[jvm-packages] add doc for xgboost4j-spark-gpu (#7779)
dmlc/xgboost: doc/jvm/xgboost4j_spark_gpu_tutorial.rst

**********************
Submit the application
**********************

Take submitting the spark job to Spark Standalone cluster as an example, and assuming your application main class
is ``Iris`` and the application jar is ``iris-1.0.0.jar``

.. code-block:: bash

  cudf_version=22.02.0
  rapids_version=22.02.0
  xgboost_version=1.6.0
  main_class=Iris
  app_jar=iris-1.0.0.jar

  spark-submit \
    --master $master\
    --packages ai.rapids:cudf:${cudf_version},com.nvidia:rapids-4-spark_2.12:${rapids_version},ml.dmlc:xgboost4j-gpu_2.12:${xgboost_version},ml.dmlc:xgboost4j-spark-gpu_2.12:${xgboost_version}\
    --conf spark.executor.cores=12 \
    --conf spark.task.cpus=1 \
    --conf spark.executor.resource.gpu.amount=1 \
    --conf spark.task.resource.gpu.amount=0.08 \
    --conf spark.rapids.sql.enabled=true \
    --conf spark.rapids.sql.csv.read.double.enabled=true \
    --conf spark.rapids.sql.hasNans=false\
    --conf spark.sql.adaptive.enabled=false \
    --conf spark.rapids.sql.explain=ALL \
    --conf spark.plugins=com.nvidia.spark.SQLPlugin \
    --class ${main_class}\

>      ${app_jar}\

Reviewer: nit: add a space between the parameter and "\".
And why you need the last "\" if there's no more parameter?

	|         5.7|        2.5|               5.0|                2.0|    2|[-2.1498908996582...|[0.00280966912396...|       2.0|
	|         5.7|        3.0|               4.2|                1.2|    1|[-2.1498908996582...|[0.00643939292058...|       1.0|
	+------------+-----------+------------------+-------------------+-----+--------------------+--------------------+----------+

**********************
Submit the application
**********************

Take submitting the spark job to Spark Standalone cluster as an example, and assuming your application main class
is ``Iris`` and the application jar is ``iris-1.0.0.jar``

.. code-block:: bash

  cudf_version=22.02.0
  rapids_version=22.02.0
  xgboost_version=1.6.0
  main_class=Iris
  app_jar=iris-1.0.0.jar

  spark-submit \
    --master $master \
    --packages ai.rapids:cudf:${cudf_version},com.nvidia:rapids-4-spark_2.12:${rapids_version},ml.dmlc:xgboost4j-gpu_2.12:${xgboost_version},ml.dmlc:xgboost4j-spark-gpu_2.12:${xgboost_version} \
    --conf spark.executor.cores=12 \
    --conf spark.task.cpus=1 \
    --conf spark.executor.resource.gpu.amount=1 \
    --conf spark.task.resource.gpu.amount=0.08 \

>     --conf spark.rapids.sql.enabled=true \

Reviewer: Some parameters are not necessary, I think.
Like spark.rapids.sql.enabled, spark.sql.adaptive.enabled, spark.rapids.sql.explain.

#############################################
XGBoost4J-Spark-GPU Tutorial (version 1.6.0+)
#############################################

**XGBoost4J-Spark-GPU** is a project aiming to accelerate XGBoost distributed training on Spark from

> end to end by GPUs by leveraging the ``Spark-Rapids``.

Reviewer 2: ```suggestion
end to end with GPUs by leveraging the ``Spark-Rapids`` project.
```

> ``Spark-Rapids`` is the RAPIDS Accelerator for Apache Spark leverages GPUs to accelerate processing

Reviewer 2: ```suggestion
``Spark-Rapids`` is the RAPIDS Accelerator for Apache Spark, which leverages GPUs to accelerate processing
```

via the `RAPIDS libraries <https://rapids.ai/>`_, for more details, please refer to the
`spark-rapids <https://nvidia.github.io/spark-rapids/>`_ page.

This tutorial will show you how to use **XGBoost4J-Spark-GPU**.

************************************************
Build an ML Application with XGBoost4J-Spark-GPU
************************************************

> Refer to XGBoost4J-Spark-GPU related Dependencies

Reviewer 2: This is not a title.

> Refer to XGBoost4J-Spark-GPU related Dependencies

Reviewer 2: ```suggestion
Refer to XGBoost4J-Spark-GPU for the related Dependencies
```

  val spark = SparkSession.builder().getOrCreate()

  val labelName = "class"
  val schema = new StructType(Array(
      StructField("sepal length", DoubleType, true),
      StructField("sepal width", DoubleType, true),
      StructField("petal length", DoubleType, true),
      StructField("petal width", DoubleType, true),
      StructField(labelName, StringType, true)))

  val xgbInput = spark.read.option("header", "false")
      .schema(schema)
      .csv(dataPath)

At the first line, we create an instance of `SparkSession <https://spark.apache.org/docs/latest/sql-getting-started.html#starting-point-sparksession>`_
which is the entry of any Spark program working with DataFrame. The ``schema`` variable
defines the schema of DataFrame wrapping Iris data. With this explicitly set schema, we
can define the columns' name as well as their types; otherwise the column name would be
the default ones derived by Spark, such as ``_col0``, etc. Finally, we can use Spark's
built-in csv reader to load Iris csv file as a DataFrame named ``xgbInput``.

Spark also contains many built-in readers for other format. eg ORC, Parquet, Avro, Json.

Transform Raw Iris Dataset
--------------------------

> To make Iris dataset be recognizable to XGBoost, we need to transform String-typed

Reviewer 2: ```suggestion
To make Iris dataset be recognizable to XGBoost, we need to encode String-typed
```

label, i.e. "class", to Double-typed label.

> To convert String-typed label to Double, we can use Spark's built-in feature transformer

Reviewer 2: ```suggestion
One way to convert the String-typed label to Double is to use Spark's built-in feature transformer
```

`StringIndexer <https://spark.apache.org/docs/2.3.1/api/scala/index.html#org.apache.spark.ml.feature.StringIndexer>`_.

> but it has not been accelerated by by Spark-Rapids, which means it will be fallen back

Reviewer 2: ```suggestion
but it has not been accelerated by Spark-Rapids yet, which means it will fall back
```

> to CPU to run and cause performance issue. Instead, we use an alternative way to acheive
> this goal by the following code
> 

Reviewer 2: ```suggestion
to CPU. As a result, we roll out our own pre-processing step here for greater efficiency:

```

.. code-block:: scala

  import org.apache.spark.sql.expressions.Window
  import org.apache.spark.sql.functions._

  val spec = Window.orderBy(labelName)
  val Array(train, test) = xgbInput
      .withColumn("tmpClassName", dense_rank().over(spec) - 1)
      .drop(labelName)
      .withColumnRenamed("tmpClassName", labelName)
      .randomSplit(Array(0.7, 0.3), seed = 1)

  train.show(5)

.. code-block:: none

	+------------+-----------+------------+-----------+-----+
	|sepal length|sepal width|petal length|petal width|class|
	+------------+-----------+------------+-----------+-----+
	|         4.3|        3.0|         1.1|        0.1|    0|
	|         4.4|        2.9|         1.4|        0.2|    0|
	|         4.4|        3.0|         1.3|        0.2|    0|
	|         4.4|        3.2|         1.3|        0.2|    0|
	|         4.6|        3.2|         1.4|        0.2|    0|
	+------------+-----------+------------+-----------+-----+

With window operations, we have mapped string column of labels to label indices.

Training
========

> XGBoost supports both regression and classification. While we use Iris dataset in

Reviewer 2: ```suggestion
The GPU version of XGBoost-Spark supports both regression and classification models. Although we use the Iris dataset in
```

> this tutorial to show how we use XGBoost/XGBoost4J-Spark to resolve a multi-classes
> classification problem, the usage in Regression is very similar to classification.

Reviewer 2: ```suggestion
this tutorial to demonstrate training XGBoost/XGBoost4J-Spark multi-classes
classification model, the concepts introduced should be readily applicable to regression problems as well.
```

To train a XGBoost model for classification, we need to claim a XGBoostClassifier first:

.. code-block:: scala

  import ml.dmlc.xgboost4j.scala.spark.XGBoostClassifier
  val xgbParam = Map(
      "objective" -> "multi:softprob",
      "num_class" -> 3,
      "num_round" -> 100,
      "tree_method" -> "gpu_hist",
      "num_workers" -> 1)

  val featuresNames = schema.fieldNames.filter(name => name != labelName)

  val xgbClassifier = new XGBoostClassifier(xgbParam)
      .setFeaturesCol(featuresNames)
      .setLabelCol(labelName)

> The available parameters for training a XGBoost model can be found in :doc:`here </parameter>`.
> Same with XGBoost4J-Spark, XGBoost4J-Spark-GPU also supports not only the default set of parameters
> but also the camel-case variant of these parameters to keep consistent with Spark's MLLIB parameters.

Reviewer 2: ```suggestion
The available parameters for training a XGBoost model can be found in :doc:`here </parameter>`.
Similar to the XGBoost4J-Spark package, in addition to the default set of parameters, XGBoost4J-Spark-GPU also supports the camel-case variant of these parameters to be consistent with Spark's MLLIB naming convention.
```

Specifically, each parameter in :doc:`this page </parameter>` has its equivalent form in
XGBoost4J-Spark-GPU with camel case. For example, to set ``max_depth`` for each tree, you can pass
parameter just like what we did in the above code snippet (as ``max_depth`` wrapped in a Map), or
you can do it through setters in XGBoostClassifer:

.. code-block:: scala

  val xgbClassifier = new XGBoostClassifier(xgbParam)
      .setFeaturesCol(featuresNames)
      .setLabelCol(labelName)
  xgbClassifier.setMaxDepth(2)

.. note::

>   XGBoost4J-Spark first needs to assemble the numeric feature columns into one column with VectorUDF
>   type by VectorAssembler, but XGBoost4J-Spark-GPU does not need such transformation, it accepts an
>   array of feature column names by ``setFeaturesCol(featuresNames)``.

Reviewer 2: ```suggestion
  In contrast to the XGBoost4J-Spark package, which needs to first assemble the numeric feature columns into one column with VectorUDF type by VectorAssembler, the XGBoost4J-Spark-GPU does not require such transformation, it accepts an array of feature column names by ``setFeaturesCol(featuresNames)``.
```

After we set XGBoostClassifier parameters and feature/label columns, we can build a transformer,
XGBoostClassificationModel by fitting XGBoostClassifier with the input DataFrame. This ``fit``

> operation is essentially the training process and the generated model can then be used in prediction.

Reviewer 2: ```suggestion
operation is essentially the training process and the generated model can then be used in other tasks like prediction.
```

.. code-block:: scala

  val xgbClassificationModel = xgbClassifier.fit(xgbInput)

Prediction
==========

When we get a model, either XGBoostClassificationModel or XGBoostRegressionModel, it takes a DataFrame,
read the column containing feature vectors, predict for each feature vector, and output a new DataFrame
with the following columns by default:

* XGBoostClassificationModel will output margins (``rawPredictionCol``), probabilities(``probabilityCol``) and the eventual prediction labels (``predictionCol``) for each possible label.
* XGBoostRegressionModel will output prediction label(``predictionCol``).

.. code-block:: scala

  val xgbClassificationModel = xgbClassifier.fit(train)
  val results = xgbClassificationModel.transform(test)
  results.show()

> With the above code snippet, we get a result DataFrame, result containing margin, probability for each class and the prediction for each instance

Reviewer 2: ```suggestion
With the above code snippet, we get a DataFrame as result, which contains the margin, probability for each class, and the prediction for each instance
```

	+------------+-----------+------------------+-------------------+-----+--------------------+--------------------+----------+

**********************
Submit the application
**********************

Take submitting the spark job to Spark Standalone cluster as an example

.. code-block:: bash

  cudf_version=22.02.0
  rapids_version=22.02.0
  xgboost_version=1.6.0

  spark-submit \
    --master $master\
    --packages ai.rapids:cudf:${cudf_version},com.nvidia:rapids-4-spark_2.12:${rapids_version},ml.dmlc:xgboost4j-gpu_2.12:${xgboost_version},ml.dmlc:xgboost4j-spark-gpu_2.12:${xgboost_version}\
    --conf spark.executor.cores=12 \
    --conf spark.task.cpus=1 \
    --conf spark.executor.resource.gpu.amount=1 \
    --conf spark.task.resource.gpu.amount=0.08 \
    --conf spark.rapids.sql.enabled=true \
    --conf spark.rapids.sql.csv.read.double.enabled=true \
    --conf spark.rapids.sql.hasNans=false\
    --conf spark.sql.adaptive.enabled=false \
    --conf spark.rapids.sql.explain=ALL \
    --conf spark.plugins=com.nvidia.spark.SQLPlugin \

* First, we need to specify the spark-rapids, cudf, xgboost4j-gpu, xgboost4j-spark-gpu packages by ``--packages``

> * Second, ``spark-rapids`` is a Spark plugin, so we need to configure it by ``spark.plugins=com.nvidia.spark.SQLPlugin``

Reviewer 2: ```suggestion
* Second, ``spark-rapids`` is a Spark plugin, so we need to configure it by specifying ``spark.plugins=com.nvidia.spark.SQLPlugin``
```

For ``spark.executor.resource.gpu.amount` and `spark.task.resource.gpu.amount``, which is related to GPU scheduling, please refer
to `Spark GPU Scheduling Overview <https://nvidia.github.io/spark-rapids/Getting-Started/#spark-gpu-scheduling-overview>`_

> when enabling `spark.rapids.sql.explain=ALL`, we can get some useful information about whether some spark physical plans can be

Reviewer 2: ```suggestion
when `spark.rapids.sql.explain=ALL` is enabled, we can get some useful information about whether some spark physical plans can be
```

	|         5.7|        3.0|               4.2|                1.2|    1|[-2.1498908996582...|[0.00643939292058...|       1.0|
	+------------+-----------+------------------+-------------------+-----+--------------------+--------------------+----------+

**********************
Submit the application
**********************

Take submitting the spark job to Spark Standalone cluster as an example

.. code-block:: bash

  cudf_version=22.02.0
  rapids_version=22.02.0
  xgboost_version=1.6.0

  spark-submit \
    --master $master\
    --packages ai.rapids:cudf:${cudf_version},com.nvidia:rapids-4-spark_2.12:${rapids_version},ml.dmlc:xgboost4j-gpu_2.12:${xgboost_version},ml.dmlc:xgboost4j-spark-gpu_2.12:${xgboost_version}\
    --conf spark.executor.cores=12 \
    --conf spark.task.cpus=1 \
    --conf spark.executor.resource.gpu.amount=1 \
    --conf spark.task.resource.gpu.amount=0.08 \
    --conf spark.rapids.sql.enabled=true \
    --conf spark.rapids.sql.csv.read.double.enabled=true \
    --conf spark.rapids.sql.hasNans=false\
    --conf spark.sql.adaptive.enabled=false \
    --conf spark.rapids.sql.explain=ALL \

>     --conf spark.plugins=com.nvidia.spark.SQLPlugin \

Reviewer: Shall you provide a complete command here?
The end "\" looks strange.

Author: Done

* First, we need to specify the spark-rapids, cudf, xgboost4j-gpu, xgboost4j-spark-gpu packages by ``--packages``
* Second, ``spark-rapids`` is a Spark plugin, so we need to configure it by specifying ``spark.plugins=com.nvidia.spark.SQLPlugin``

For ``spark.executor.resource.gpu.amount` and `spark.task.resource.gpu.amount``, which is related to GPU scheduling, please refer
to `Spark GPU Scheduling Overview <https://nvidia.github.io/spark-rapids/Getting-Started/#spark-gpu-scheduling-overview>`_

> when `spark.rapids.sql.explain=ALL` is enabled, we can get some useful information about whether some spark physical plans can be

Reviewer: Below part from line 241 to line 283 is not related to xgboost, i think.
Shall you remove it?

Author: Done
